cdap-app-templates/cdap-program-report/src/main/scala/co/cask/cdap/report/ReportGenerationHelper.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.report
+
+import co.cask.cdap.report.proto.ReportGenerationRequest
+import co.cask.cdap.report.util.Constants
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{SQLContext, SparkSession}
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConversions._
+
+class ReportGenerationHelper(private val spark: SparkSession) {
+
+  def this(sc: SparkContext) {
+    this(new SQLContext(sc).sparkSession)
+  }
+
+  def generateReport(request: ReportGenerationRequest, inputPaths: java.util.List[String], outputPath: String): Long = {
+    import ReportGenerationHelper._
+    import spark.implicits._
+    val df = spark.read.format("com.databricks.spark.avro").load(asScalaBuffer(inputPaths): _*)
+    // Group the program run meta records by program run Id's and aggregate grouped records into a column
+    // with data type Record. The aggregated DataFrame aggDf will have two columns: "run" and "record"
+    val aggCol = new RecordAgg().toColumn.alias(RECORD_COL).as[Record]
+    // TODO: configure partitions. The default number of partitions is 200
+    var aggDf = df.groupBy(Constants.RUN).agg(aggCol)
+    // Construct a set of fields to be included in the final report with required fields and fields from the request
+    val reportFields = collection.mutable.LinkedHashSet(REQUIRED_FIELDS: _*)
+    if (Option(request.getFields).isDefined) reportFields ++= request.getFields
+    LOG.debug("reportFields={}", reportFields)
+    // Construct a set of additional fields to be included as columns in the aggregated DataFrame
+    // with fields used for filtering and sorting
+    val additionalFields = collection.mutable.LinkedHashSet(Constants.START, Constants.END)
```

Comment:
Similar to above, you should be able to do it in one line.

Reply:
This is a bit different from the case above 
```scala
  if (Option(request.getFilters).isDefined) {
     request.getFilters.foreach(f => additionalFields.add(f.getFieldName))
    }
```
There is an additional step of transforming the `f: Filter` to a String `f.getFieldName)` 
```scala
additionalFields ++= Option(request.getFilters).getOrElse(Nil)
``` 
won't work since their types don't match, and it seems that I cannot call map on ```Option(request.getFilters).getOrElse(Nil)``` either

Reply:
You can easily transform an `Seq` with the `map` method:

```
val additionalFields = mutable.LinkedHashSet(Constants.START, Constants.END) ++= Option(request.getFilters).getOrElse(Nil).map(_.getFieldName)
```